AssertionError when wrongly mapping a 2D array

I got the following error when I want to compile the attached spec:
```
[error] (run-main-1) java.lang.AssertionError: assertion failed
[error] java.lang.AssertionError: assertion failed
[error] 	at scala.Predef$.assert(Predef.scala:265)
[error] 	at pgo.trans.MPCalPCalCodegenPass$.$anonfun$apply$10(MPCalPCalCodegenPass.scala:204)
[error] 	at scala.Option.flatMap(Option.scala:283)
[error] 	at pgo.trans.MPCalPCalCodegenPass$.pgo$trans$MPCalPCalCodegenPass$$applyMappingMacros$1(MPCalPCalCodegenPass.scala:199)
[error] 	at pgo.trans.MPCalPCalCodegenPass$.$anonfun$apply$18(MPCalPCalCodegenPass.scala:238)
[error] 	at scala.collection.immutable.List.flatMap(List.scala:293)
[error] 	at pgo.trans.MPCalPCalCodegenPass$.pgo$trans$MPCalPCalCodegenPass$$applyMappingMacros$1(MPCalPCalCodegenPass.scala:238)
[error] 	at pgo.trans.MPCalPCalCodegenPass$$anonfun$5.$anonfun$applyOrElse$12(MPCalPCalCodegenPass.scala:373)
[error] 	at scala.collection.Iterator$$anon$10.nextCur(Iterator.scala:585)
[error] 	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:599)
[error] 	at scala.collection.immutable.List.prependedAll(List.scala:152)
[error] 	at scala.collection.IterableOnceOps.toList(IterableOnce.scala:1251)
[error] 	at scala.collection.IterableOnceOps.toList$(IterableOnce.scala:1251)
[error] 	at scala.collection.AbstractIterable.toList(Iterable.scala:919)
[error] 	at pgo.trans.MPCalPCalCodegenPass$$anonfun$5.applyOrElse(MPCalPCalCodegenPass.scala:373)
[error] 	at pgo.trans.MPCalPCalCodegenPass$$anonfun$5.applyOrElse(MPCalPCalCodegenPass.scala:327)
[error] 	at pgo.model.Rewritable$TopDownFirstStrategy$.execute(Rewritable.scala:174)
[error] 	at pgo.model.Rewritable$TopDownFirstStrategy$.$anonfun$execute$3(Rewritable.scala:181)
[error] 	at scala.collection.immutable.List.map(List.scala:246)
[error] 	at scala.collection.immutable.List.map(List.scala:79)
[error] 	at pgo.model.Rewritable$.transformSubNodes(Rewritable.scala:160)
[error] 	at pgo.model.Rewritable$TopDownFirstStrategy$.execute(Rewritable.scala:181)
[error] 	at pgo.model.Rewritable$TopDownFirstStrategy$.$anonfun$execute$2(Rewritable.scala:176)
[error] 	at scala.collection.Iterator$$anon$9.next(Iterator.scala:575)
[error] 	at scala.collection.immutable.List.prependedAll(List.scala:156)
[error] 	at scala.collection.immutable.List$.from(List.scala:651)
[error] 	at scala.collection.immutable.List$.from(List.scala:648)
[error] 	at scala.collection.SeqFactory$Delegate.from(Factory.scala:306)
[error] 	at scala.collection.immutable.Seq$.from(Seq.scala:42)
[error] 	at scala.collection.IterableOnceOps.toSeq(IterableOnce.scala:1263)
[error] 	at scala.collection.IterableOnceOps.toSeq$(IterableOnce.scala:1263)
[error] 	at scala.collection.AbstractIterator.toSeq(Iterator.scala:1288)
[error] 	at pgo.model.Rewritable.withChildren(Rewritable.scala:101)
[error] 	at pgo.model.Rewritable.withChildren$(Rewritable.scala:100)
[error] 	at pgo.model.mpcal.MPCalNode.withChildren(AST.scala:7)
[error] 	at pgo.model.Rewritable.mapChildren(Rewritable.scala:92)
[error] 	at pgo.model.Rewritable.mapChildren$(Rewritable.scala:45)
[error] 	at pgo.model.mpcal.MPCalNode.mapChildren(AST.scala:7)
[error] 	at pgo.model.Rewritable$TopDownFirstStrategy$.execute(Rewritable.scala:176)
[error] 	at pgo.model.Rewritable.rewrite(Rewritable.scala:113)
[error] 	at pgo.model.Rewritable.rewrite$(Rewritable.scala:112)
[error] 	at pgo.model.mpcal.MPCalNode.rewrite(AST.scala:7)
[error] 	at pgo.trans.MPCalPCalCodegenPass$.apply(MPCalPCalCodegenPass.scala:327)
[error] 	at pgo.PGo$.run(PGo.scala:134)
[error] 	at pgo.PGo$.main(PGo.scala:224)
[error] 	at pgo.PGo.main(PGo.scala)
[error] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error] 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
[error] stack trace is suppressed; run last Compile / bgRun for the full output
[error] Nonzero exit code: 1
[error] (Compile / run) Nonzero exit code: 1
[error] Total time: 3 s, completed Oct. 11, 2021, 10:31:52 a.m.
```

Attached spec: https://gist.github.com/shayanh/f746283780fa86e98f02f5590d3da0fc

--- a/src/pgo/parser/PCalParser.scala
+++ b/src/pgo/parser/PCalParser.scala
@@ -86,10 +86,16 @@ trait PCalParser extends TLAParser {
 
   def pcalLhs(implicit ctx: PCalParserContext): Parser[PCalAssignmentLhs] = {
       def rec(lhs: PCalAssignmentLhs): Parser[PCalAssignmentLhs] =
-        opt(ws ~> withSourceLocation {
-          "." ~> ws ~> tlaIdentifierExpr ^^ (id => PCalAssignmentLhsProjection(lhs, List(TLAString(id.id).setSourceLocation(id.sourceLocation)))) |
-          "[" ~> ws ~> rep1sep(tlaExpression, ws ~> "," ~> ws) <~ ws <~ "]" ^^ (PCalAssignmentLhsProjection(lhs, _))
-        }.flatMap(rec)).map(_.getOrElse(lhs))
+        opt {
+          ws ~> (querySourceLocation {
+            "." ~> ws ~> tlaIdentifierExpr ^^ (id => PCalAssignmentLhsProjection(lhs, List(TLAString(id.id).setSourceLocation(id.sourceLocation)))) |
+              "[" ~> ws ~> rep1sep(tlaExpression, ws ~> "," ~> ws) <~ ws <~ "]" ^^ (PCalAssignmentLhsProjection(lhs, _))
+          } ^^ {
+            case (loc, proj) => proj.setSourceLocation(lhs.sourceLocation ++ loc) // ensure the projection's position includes LHS
+          })
+            .flatMap(rec)
+        }
+          .map(_.getOrElse(lhs))
 
       pcalLhsId.flatMap(rec)
     }

--- a/src/pgo/trans/MPCalPCalCodegenPass.scala
+++ b/src/pgo/trans/MPCalPCalCodegenPass.scala
@@ -199,8 +199,8 @@ object MPCalPCalCodegenPass {
             findRef(lhs).flatMap {
               case ident if mappingsMap.contains(ById(ident.refersTo)) =>
                 val (mappingCount, mapping) = mappingsMap(ById(ident.refersTo))
-                // maybe implement depth > mappingCount in the future, but it may be more confusion than it's worth
-                // for now, assert that it's not the case, to avoid very confusing mis-compilations
+                // this should always be the case; it would be too confusing otherwise, as documented in
+                // test/files/general/pcalgen/MappingMacroRWExpansion.tla
                 assert(mappingCount == findLhsDepth(lhs))
                 val convertedLhs = convertLhs(lhs)
                 val valueBind = PCalVariableDeclarationValue(TLAIdentifier(nameCleaner.cleanName("value")), rhs)

--- a/src/pgo/trans/MPCalSemanticCheckPass.scala
+++ b/src/pgo/trans/MPCalSemanticCheckPass.scala
@@ -44,6 +44,11 @@ object MPCalSemanticCheckPass {
         defn.sourceLocation.longDescription
       }\n does not match")
 
+    final case class MPCalReadWriteAssignmentForbidden(usage: SourceLocatable, defn: SourceLocatable) extends Error(
+      usage.sourceLocation, d"assignment implies both a read from and a write to the state variable defined at ${
+        defn.sourceLocation.longDescription
+      }\n if this is intended, use a with statement to explicitly show the separate read and write operations")
+
     final case class PClArityMismatch(usage: SourceLocatable, defn: SourceLocatable) extends Error(
       usage.sourceLocation, d"call arity at ${
         defn.sourceLocation.longDescription
@@ -342,8 +347,15 @@ object MPCalSemanticCheckPass {
               findDefn(lhs).foreach { defn =>
                 val mappingCount = findMappingCount(lhs)
                 val mappingCountP = defn.mappingCount
-                if(mappingCountP > mappingCount) {
-                  errors += SemanticError.MPCalKindMismatchError(usage = lhs, defn = defn)
+                // the mapping count must match exactly:
+                // - too few, and there is no reasonable behaviour
+                // - too many, and it implies a combined read-write, which would be confusing
+                if(mappingCountP != mappingCount) {
+                  if(mappingCountP < mappingCount) {
+                    errors += SemanticError.MPCalReadWriteAssignmentForbidden(usage = lhs, defn = defn)
+                  } else {
+                    errors += SemanticError.MPCalKindMismatchError(usage = lhs, defn = defn)
+                  }
                 }
               }
             case ref@MPCalRefExpr(_, mappingCount) =>

--- a/test/files/general/MPCalKindMatching.tla
+++ b/test/files/general/MPCalKindMatching.tla
@@ -9,7 +9,7 @@ EXTENDS Sequences, FiniteSets, Integers
         l4: call Proc2(ref a[_][_]);
         l5: a[2] := 3;
         l3: (*:: expectedError: MPCalKindMismatchError *) a := 3;
-        l6: a[5][6] := 3;
+        l6: (*:: expectedError: MPCalReadWriteAssignmentForbidden *) a[5][6] := 3;
         l7: x := (*:: expectedError: MPCalKindMismatchError *) a;
         l8: x := a[3];
         l9: x := a[(*:: expectedError: MPCalKindMismatchError *) a];

--- a/test/files/pcalgen/MappingMacroRWExpansion.tla
+++ b/test/files/pcalgen/MappingMacroRWExpansion.tla
@@ -1,0 +1,43 @@
+------------------------------- MODULE func -------------------------------
+
+EXTENDS Naturals, Sequences, TLC, FiniteSets
+
+(********************
+
+--mpcal func {
+
+    mapping macro MP {
+        read {
+            yield $variable;
+        }
+
+        write {
+            yield $value;
+        }
+    }
+
+    archetype ANode(ref arr[_]) {
+    lbl1:
+        (* this has the "wrong" number of indices, and implies:
+            - read arr[1] (call it x)
+            - write [x EXCEPT ![2] = "a"] to arr[1]
+
+          this is hard to keep track of, so it's forbidden instead *)
+        (*:: expectedError: MPCalReadWriteAssignmentForbidden *) arr[1][2] := "a";
+
+    lbl2:
+        print arr[1][2];
+    }
+
+    variable
+        arr = [d1 \in {1} |-> [d2 \in {2} |-> ""]];
+
+    fair process (Node \in {1}) == instance ANode(ref arr[_])
+        mapping arr[_] via MP;
+}
+
+********************)
+
+\* BEGIN TRANSLATION
+
+=============================================================================

